topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,234 @@
+	XMLDesc, _ := Vm.GetXMLDesc(0)
+	nodes := helpers.FindInterfacesVMConnectedThrough(XMLDesc)
+	logging.GetLogger().Debugf("Register virtual machine: %s", VmName)
+
+	VmID, _ := Vm.GetID()
+
+	logging.GetLogger().Debugf("Register VM: %s", VmName)
+	state_stringified := probe.getVirtualMachineStatus(Vm)
+	metadata := &graph.Metadata{
+		"Type":    MANAGER,
+		"Manager": MANAGER,
+		"Name":    VmName,
+		"LibVirt": map[string]interface{}{
+			"ID":   VmID,
+			"Name": VmName,
+		},
+		"State": state_stringified,
+	}
+	probe.GraphRealtimeHandler.AddNode(VmName, metadata, nodes)
+	logging.GetLogger().Debugf("Finalized registration of virtual machine: %s", VmName)
+	return nil
+}
+
+func (probe *LibVirtProbe) handleLibVirtEvent(Connection *libvirt.Connect, Domain *libvirt.Domain, e *libvirt.DomainEventLifecycle) {
+	VmName, _ := Domain.GetName()
+	logging.GetLogger().Debugf("got libvirt event: %d - %s", e.Event, VmName)
+	switch e.Event {
+	case libvirt.DOMAIN_EVENT_UNDEFINED:
+		probe.unregisterVM(Domain)
+		break
+	case libvirt.DOMAIN_EVENT_STARTED:
+		probe.changeStatusOfVm(Domain)
+		probe.registerVM(Domain)
+		break
+	case libvirt.DOMAIN_EVENT_STOPPED:
+		probe.changeStatusOfVm(Domain)
+		break
+	case libvirt.DOMAIN_EVENT_SHUTDOWN:
+		probe.changeStatusOfVm(Domain)
+		break
+	case libvirt.DOMAIN_EVENT_DEFINED:
+		probe.registerVM(Domain)
+		break
+	}
+	return
+}
+
+func (probe *LibVirtProbe) isLibvirtCallbackConnected() bool {
+	return probe.libvirtCallbackId != -1
+}
+
+func (probe *LibVirtProbe) periodicLibvirtChecker() {
+	for true {
+		if !probe.isLibvirtCallbackConnected() {
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		libvirt.EventRunDefaultImpl()
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (probe *LibVirtProbe) Connect() error {
+	var err error
+
+	logging.GetLogger().Debugf("Connecting to virsh daemon: %s", probe.url)
+	// explanation why we need this stuff is here:
+	// https://libvirt.org/html/libvirt-libvirt-event.html#virEventRegisterImpl
+	// and here https://libvirt.org/html/libvirt-libvirt-event.html#virEventRunDefaultImpl
+	libvirt.EventRegisterDefaultImpl()
+	probe.Client, err = libvirt.NewConnectReadOnly(probe.url)
+	if err != nil {
+		logging.GetLogger().Errorf("Failed to create client to libvirt daemon: %s", err.Error())
+		return err
+	}
+	// register libvirt callback
+	libvirtCallbackId, err := probe.Client.DomainEventLifecycleRegister(nil, probe.handleLibVirtEvent)
+	if err != nil {
+		logging.GetLogger().Errorf("Not able to connect to libvirt event loop. Error: %s", err.Error())
+		return err
+	} else {
+		probe.libvirtCallbackId = libvirtCallbackId
+		logging.GetLogger().Infof("Subscribed to libvirt events. Callback id: %d", libvirtCallbackId)
+	}
+	_, cancel := context.WithCancel(context.Background())
+
+	probe.Cancel = cancel
+	probe.Wg.Add(1)
+
+	probe.Connected.Store(true)
+	defer probe.Connected.Store(false)
+
+	go func() {
+		defer probe.Wg.Done()
+		defer probe.periodicLibvirtChecker()
```

Comment:
I am surprised that you had no problem with the loop containing EventRunDefaultImpl not active when the libvirt connection is created. Moving it from the end of the startProbe code to the begining made my code work and it should definitely be active before the initial list of domains is performed.

Reply:
To be honest I am not sure, I tested adding/removing vms in realtime, the graph updated, etc
So, maybe you are right, but I've not seen any real issue with it.